### DISCUSSION TITLE
Configuration format

### DISCUSSION THREAD

**Discussion by rlemaitre** [rlemaitre/pillars#187] [General]
I always wonder what is the best format for configuring a backend service.

In my opinion, a good configuration must:

- be easily readable by humans (and the program obviously 😄 )
- be easily writable by humans
- have logical defaults
- be not too verbose
- fail as soon as possible if the configuration is faulty (missing mandatory params for example)
- support
  - basic types (String – single and multi-lines, Integers, floats, date and time, booleans, duration, url)
  - arrays
  - comments
  - dictionaries (aka `Map`)
  - import from file and/or url
  - validation against a schema
  - override configuration file parameters by environment variables and command-line arguments (CLI > Env Var > file)
  
Ideally, it would allow being stored in a VCS, allow typo detection (by signalling non handled params at startup for example), support parameter depreciation, and support strong typing. And I can also imagine use cases where computation could be useful (this parameter value is the half of another one for example)

So, besides classical formats (TOML, YAML, JSON), we could also imagine using some programming languages to do this. I'm thinking about DHAL or even Scala (loaded via Ammonite at startup).

What do you think about configuration? What would be your preferred way to configure a Pillars app?

**Comment by jnicoulaud-ledger** [comment]
I would add to your criteria:
 - Have at least basic variable interpolation
 - Can default to env vars
 - Has basic syntax highlighting/check/formatter in most IDEs

**Comment by rlemaitre** [comment]
@onouguier-ledger told me about [PKL](https://github.com/apple/pkl). It looks promising. I'll try to make a PoC

**Comment by greshny** [comment]
What do you think about [ciris](https://github.com/vlovgr/ciris)?

**Reply by rlemaitre** [reply]
I think it's a great lib for developers, but, as it reads static configuration files, it doesn't answer to several problems such as variable or even functions declaration, schema enforcement, ...
